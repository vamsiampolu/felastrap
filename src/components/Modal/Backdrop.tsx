import { useFela } from "react-fela";
import { DivProps } from "../common";
import { CSSProperties } from "react";

const backdropRule = () => ({
  backgroundColor: "#000",
  opacity: 0,
  height: "100vh",
  width: "100vw",
  left: 0,
  position: "fixed",
  top: 0,
  zIndex: 1050,
});

export type BackdropProps = DivProps & {
  isOpen?: boolean;
  style?: CSSProperties | null;
};

export const Backdrop = ({ className = "", style }: BackdropProps) => {
  const { css } = useFela();
  return (
    <div
      data-testid="modal-backdrop"
      style={style}
      className={`${css(backdropRule)} bs-modal-backdrop ${className}`.trim()}
    ></div>
  );
};
