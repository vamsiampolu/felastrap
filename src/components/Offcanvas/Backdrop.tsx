import { CSSProperties } from "react";
import { useFela } from "react-fela";

const rule = ({ isOpen = false }) => ({
  position: "fixed",
  opacity: isOpen ? 0.5 : 0,
  top: 0,
  left: 0,
  zIndex: 1040,
  width: "100vw",
  height: "100vh",
  backgroundColor: "#000",
});

export type BackdropProps = { isOpen?: boolean; style: CSSProperties };

export const Backdrop = ({ isOpen = false, style }: BackdropProps) => {
  const { css } = useFela({ isOpen });
  const cx = `${css(rule)} offcanvas-backdrop`;
  return <div data-testid="offcanvas-backdrop" style={style} className={cx} />;
};
