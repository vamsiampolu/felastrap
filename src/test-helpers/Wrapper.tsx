import type { IRenderer } from "fela";
import { type PropsWithChildren } from "react";
import { RendererProvider, ThemeProvider } from "react-fela";

const Wrapper = ({
  renderer,
  theme = {},
  children,
}: PropsWithChildren<{ renderer: IRenderer; theme?: object }>) => (
  <ThemeProvider theme={theme}>
    <RendererProvider renderer={renderer}>{children}</RendererProvider>;
  </ThemeProvider>
);

export { Wrapper };
