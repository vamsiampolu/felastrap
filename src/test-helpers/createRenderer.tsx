import expandShorthand from "fela-plugin-expand-shorthand";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";
import prefixer from "fela-plugin-prefixer";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import { createRenderer as createFelaRenderer, type IRenderer } from "fela";
import { lightTheme } from "../reboot/lightTheme";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  const enhancers = [sortClassNames(), sortMediaQueryMobileFirst()];
  const plugins = [
    prefixer(),
    placeholderPrefixer(),
    expandShorthand(autoMergeLonghands),
  ];

  return createFelaRenderer({
    plugins,
    enhancers,
  });
};

export const cssReboot = (renderer: IRenderer) => {
  renderer.renderStatic(lightTheme);
};
