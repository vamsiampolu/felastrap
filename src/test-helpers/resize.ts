export const resizeWidth = (width: number) => {
  window.innerWidth = width;
  dispatchEvent(new Event("resize"));
};

export const resizeHeight = (height: number) => {
  window.innerHeight = height;
  dispatchEvent(new Event("resize"));
};
