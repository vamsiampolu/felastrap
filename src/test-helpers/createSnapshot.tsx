import { IRenderer } from "fela";
import { renderToString } from "fela-tools";
import { ReactElement, ReactNode } from "react";

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import {
  type UserEvent,
  type UserEventApi,
} from "@testing-library/user-event/dist/types/setup/setup";
import * as prettier from "prettier";
import { renderToStaticMarkup } from "react-dom/server";
import { Wrapper } from "./Wrapper";

async function formatCSS(css: string) {
  return prettier.format(css, {
    parser: "css",
    useTabs: false,
    tabWidth: 2,
  });
}

function formatHTML(str: string) {
  return prettier.format(str, { parser: "html" });
}

function renderToMarkup(component: ReactElement): string {
  return renderToStaticMarkup(component);
}

export const createSnapshot = async ({
  component,
  renderer,
  theme = {},
}: {
  component: ReactNode;
  renderer: IRenderer;
  theme?: object;
}) => {
  // reset renderer to have a clean setup
  renderer.clear();

  const element = (
    <Wrapper theme={theme} renderer={renderer}>
      {component}
    </Wrapper>
  );

  const view = renderToMarkup(element);
  const str = renderToString(renderer);

  const css = await formatCSS(str);
  const html = await formatHTML(view);

  return `${css}\n\n${html}`;
};

export type EventName = keyof UserEventApi | "focus" | "focus-visible";
export type CreateUserEventType = {
  eventName: EventName;
  component: ReactElement;
  renderer: IRenderer;
  testId: string;
  theme?: object;
  donotWrap?: boolean;
};

export type SetupEvent = {
  user: UserEvent;
  eventName: EventName;
  element: HTMLElement;
};
const setupEvent = async ({ user, eventName, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.hover(element);
      break;
    }

    case "focus": {
      element.focus();
      break;
    }

    case "focus-visible": {
      element.contentEditable = "true";
      element.focus();
      element.contentEditable = "false";
      break;
    }
  }
};

const clearEvent = async ({ eventName, user, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.unhover(element);
      break;
    }

    case "focus":
    case "focus-visible": {
      element.blur();
      break;
    }
  }
};

export const createUserEventSnapshot = async ({
  eventName,
  component,
  testId,
  renderer,
  theme = {},
  donotWrap = false,
}: CreateUserEventType) => {
  renderer.clear();

  const user = userEvent.setup();
  let wrappedComponent = component;
  if (!donotWrap) {
    wrappedComponent = (
      <Wrapper theme={theme} renderer={renderer}>
        {component}
      </Wrapper>
    );
  }

  render(wrappedComponent);

  const element = await screen.findByTestId(testId);

  setupEvent({ element, eventName, user });
  const view = await renderToMarkup(wrappedComponent);
  const utils = renderToString(renderer);
  const css = await formatCSS(utils);
  const html = await formatHTML(view);

  clearEvent({ eventName, element, user });

  return `${css}\n\n${html}`;
};
