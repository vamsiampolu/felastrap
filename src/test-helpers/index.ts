import { Wrapper } from "./Wrapper";
import { createRenderer, cssReboot } from "./createRenderer";
import { createSnapshot, createUserEventSnapshot } from "./createSnapshot";
import { resizeHeight, resizeWidth } from "./resize";

export {
  createRenderer,
  cssReboot,
  createSnapshot,
  createUserEventSnapshot,
  resizeWidth,
  resizeHeight,
  Wrapper,
};
