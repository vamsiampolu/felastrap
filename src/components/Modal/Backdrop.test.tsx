import { IRenderer } from "fela";
import { createRenderer, createSnapshot, cssReboot } from "../../test-helpers";
import { Backdrop } from "./Backdrop";

describe("components/Modal/Backdrop", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a Modal backdrop", async () => {
    const component = <Backdrop />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
