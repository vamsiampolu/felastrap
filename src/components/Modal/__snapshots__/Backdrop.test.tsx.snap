// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal/Backdrop should render a Modal backdrop 1`] = `
".a {
  -webkit-background-color: #000;
  background-color: #000;
}
.b {
  opacity: 0;
}
.c {
  height: 100vh;
}
.d {
  width: 100vw;
}
.e {
  left: 0;
}
.f {
  position: fixed;
}
.g {
  top: 0;
}
.h {
  z-index: 1050;
}


<div
  data-testid="modal-backdrop"
  class="a b c d e f g h bs-modal-backdrop"
></div>
;
"
`;
